faiss_vector_cache.go

```diff
@@ -18,92 +18,130 @@
-	// which will be used to calculate the average in the next cycle of average
-	// computation (which is average traffic for the field till now). this is
-	// used to track the per second hits to the cache entries.
-	sample uint64
-}
-
-type cacheEntry struct {
-	tracker *ewma
-	// this is used to track the live references to the cache entry,
-	// such that while we do a cleanup() and we see that the avg is below a
-	// threshold we close/cleanup only if the live refs to the cache entry is 0.
-	refs  int64
-	index *faiss.IndexImpl
-}
-
 func newVectorIndexCache() *vectorIndexCache {
 	return &vectorIndexCache{
 		cache:   make(map[uint16]*cacheEntry),
 		closeCh: make(chan struct{}),
 	}
 }
 
+type vectorIndexCache struct {
+	closeCh chan struct{}
+	m       sync.RWMutex
+	cache   map[uint16]*cacheEntry
+}
+
 func (vc *vectorIndexCache) Clear() {
 	vc.m.Lock()
 	close(vc.closeCh)
 
 	// forcing a close on all indexes to avoid memory leaks.
 	for _, entry := range vc.cache {
-		entry.closeIndex()
+		entry.close()
 	}
 	vc.cache = nil
 	vc.m.Unlock()
 }
 
-func (vc *vectorIndexCache) loadVectorIndex(fieldID uint16,
-	indexBytes []byte) (vecIndex *faiss.IndexImpl, err error) {
-	cachedIndex, present := vc.isIndexCached(fieldID)
-	if present {
-		vecIndex = cachedIndex
-		vc.incHit(fieldID)
+func (vc *vectorIndexCache) loadOrCreate(fieldID uint16, mem []byte, except *roaring.Bitmap) (
+	index *faiss.IndexImpl, vecDocIDMap map[int64]uint32, vecIDsToExclude []int64, err error) {
+	entry := vc.fetch(fieldID)
+	if entry != nil {
+		index, vecDocIDMap = entry.load()
+		if except != nil && !except.IsEmpty() {
+			for vecID, docID := range vecDocIDMap {
+				if except.Contains(docID) {
+					vecIDsToExclude = append(vecIDsToExclude, vecID)
+				}
+			}
+		}
 	} else {
-		vecIndex, err = vc.createAndCacheVectorIndex(fieldID, indexBytes)
+		index, vecDocIDMap, vecIDsToExclude, err = vc.createAndCache(fieldID, mem, except)
+	}
+	return index, vecDocIDMap, vecIDsToExclude, err
+}
+
+func (vc *vectorIndexCache) fetch(fieldID uint16) *cacheEntry {
+	vc.m.RLock()
+	defer vc.m.RUnlock()
+
+	entry, ok := vc.cache[fieldID]
+	if !ok {
+		return nil
 	}
-	vc.addRef(fieldID)
-	return vecIndex, err
+
+	entry.addRef()
+	return entry
 }
 
-func (vc *vectorIndexCache) createAndCacheVectorIndex(fieldID uint16,
-	indexBytes []byte) (*faiss.IndexImpl, error) {
+func (vc *vectorIndexCache) createAndCache(fieldID uint16, mem []byte, except *roaring.Bitmap) (
+	index *faiss.IndexImpl, vecDocIDMap map[int64]uint32, vecIDsToExclude []int64, err error) {
 	vc.m.Lock()
 	defer vc.m.Unlock()
 
 	// when there are multiple threads trying to build the index, guard redundant
 	// index creation by doing a double check and return if already created and
 	// cached.
-	entry, present := vc.cache[fieldID]
-	if present {
-		rv := entry.index
-		entry.incHit()
-		return rv, nil
+	entry, ok := vc.cache[fieldID]
+	if ok {
+		entry.addRef()
+		index, vecDocIDMap = entry.load()
```

Comment:
maybe we can keep the following snippet in a func since its repeated at a couple of places (line 60) calling it "getInvalidVecs" or something better.